src/backend/renderer.js

```diff
@@ -577,22 +577,34 @@ export function attach(
     );
   }
 
-  let pendingOperations: Uint32Array = new Uint32Array(0);
-  let pendingOperationsQueue: Array<Uint32Array> | null = [];
-
-  function addOperation(
-    newAction: Uint32Array,
-    addToStartOfQueue: boolean = false
-  ): void {
-    const oldActions = pendingOperations;
-    pendingOperations = new Uint32Array(oldActions.length + newAction.length);
+  let pendingOperations: Array<number> = [];
+  let pendingOperationsQueue: Array<Array<number>> | null = [];
+
+  let nextOperation: Array<number> = [];
+  function beginNextOperation(size: number): void {
+    nextOperation.length = size;
+  }
+  function endNextOperation(addToStartOfQueue: boolean): void {
+    if (__DEV__) {
+      for (let i = 0; i < nextOperation.length; i++) {
+        if (!Number.isInteger(nextOperation[i])) {
+          console.error(
+            'endNextOperation() was called but some values are not integers.',
+            nextOperation
+          );
+        }
+      }
+    }
+
     if (addToStartOfQueue) {
-      pendingOperations.set(newAction);
-      pendingOperations.set(oldActions, newAction.length);
+      pendingOperations.splice.apply(
+        pendingOperations,
+        [0, 0].concat(nextOperation)
+      );
```

Comment:
NB: this is still an O(n) operation which makes it O(n^2) to process n "unmount" commands, for example. A double-ended queue is probably the right data structure here if we do need to insert at both the start and end.